Refactor nordic socs/drivers to remove GPD and decouple pin retention.

bjarki-andreasen opened the pull request:
Replace highly coupled use of GPD (gpd.h) from device drivers, with power domain device drivers using zephyr's PM_DEVICE_POWER_DOMAIN infrastructure.

The gpd.h currently is used for two things:
* Manually requests power domains like fast_active_main and slow_active_main on the H20
* Manually applies pin state retention on individual pins, or pins found in a pinctrl config.

Nordic device drivers currently have to do this manually, conditionally, which is adding immense complexity to them, only relevant for certain SoCs, and which is in theory duplicate code. In reality, device drivers do their best, but since gpd is not well understood, device drivers implement it slightly differently, in some cases working by design, some cases working by accident, and in some cases working, but breaking something else.

**No more!**

Both pin retention and power domains are actually designed in a way that is entirely logical, and can be expressed using power domains and the devicetree. So, using pinctrl and power domains, we can entirely decouple devices from this complexity :)

**What does the refactor do?**

1. Rework the nrfh54h power domains to match zephyr's power domains. This means every power domain is represented as a node, not an index in a phandle. This does two things, it allows us to use zephyr's power domain framework, and it stops our misuse of the base `power-domains` property from breaking user's applications if they "dare" to use zephyrs power domain framework. See first commit.
2. Introduce the gpio pads to the devicetree. GPIO pads are the SoCs pins, which peripherals like UARTE or GPIO can be muxed to. These can be in different power domains than the peripheral connected to them, and they need to be retained if the power domain they are in may be shut down, which only happens if no peripheral, including GPIO, is active while muxed to any pad in that pad group (a pad group is 1-1 mapped to the GPIO controller, so GPIO0 will own pad group 0). See second commit.
3. Refactor pinctrl and device drivers to use the new power domains, and remove all usage of gpd.h
4. Remove gpd.h
5. Success

karstenkoenig commented on drivers/power_domain/power_domain_nrfs_gdpwr.c:
```diff
@@ -0,0 +1,227 @@
+
+	dev_data->synced = true;
+
+	if (dev_data->off) {
```
this expects domains to be by default requested after reset right? Can you leave a comment about this, otherwise it seems strange why you wouldn't just do unconditional `manager_set_domain_locked(dev_config->domain, !dev_data->off);`

bjarki-andreasen replied:
Yep, will leave a note about this :)

karstenkoenig commented on drivers/power_domain/power_domain_nrfs_gdpwr.c:
```diff
@@ -0,0 +1,227 @@
+	manager_lock();
+
+	if (dev_data->synced) {
+		ret = manager_set_domain_locked(dev_config->domain, on);
```
Why aren't you maintaining and comparing the state in domain_data.off? What happens if the power domain gets resumed multiple times or won't that happen?

karstenkoenig commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -436,12 +434,9 @@ static void uarte_disable_locked(const struct device *dev, uint32_t dis_mask)
-#endif
 	nrf_uarte_disable(get_uarte_instance(dev));
+
+	(void)pinctrl_apply_state(config->pcfg, PINCTRL_STATE_SLEEP);
```
Don't we need to move this up to avoid glitching the pin? Assuming UART tx is idle high then this might disable the uart (release pin from uart) and then drive it high here if PINCTRL_STATE_SLEEP is high

karstenkoenig reviewed (commented):
Didn't look at SWEXT changes, but the gpio port/pad power domain handling looks good.

Will this trigger more power domain requests/releases?

bjarki-andreasen-nordic committed:
drivers: power_domain: introduce nrfs gdpwr

Introduce the NRFS GDPWR (Global Domain Power Request) device
driver and devicetree binding.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
drivers: power_domain: introduce nrf_gpio_pad_group

Introduce the NRF GPIO Pad Group device driver and binding. The
pad group device represents the GPIO pads (pins), contrary to a
GPIO controller, which is one of the many devices which can be
muxed to pads in the pad group.

The pad group belong to a power domain, which is not neccesarily the
same power domain as devices being muxed to the pads, like GPIO or
UART. If no ACTIVE device is using any of the pads in the pad
group, the pad groups power domain may be SUSPENDED. Before the pad
groups power domain is SUSPENDED, pad config retention must be
enabled to prevent the pads from loosing their state. That's what
this device driver manages. Once retained, the pad configs and
outputs are locked, even when their power domain is SUSPENDED.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
soc: nordic: nrf54h: transition from gpd to zephyr pinctrl and pds

Transition nrf54h away from the soc specific gpd
(global power domain) driver which mixed power domains, pinctrl
and gpio pin retention into a non scalable solution, forcing soc
specific logic to bleed into nrf drivers.

The new solution uses zephyrs PM_DEVICE based power domains to
properly model the hardware layout of device and pin power domains,
and moves pin retention logic out of drivers into pinctrl and
gpio, which are the components which manage pins (pads).

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
soc: nrf54h: remove deprecated gpd (global power domain) driver

Remove the deprecated GPD (Global Power Domain) driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

masz-nordic commented on drivers/serial/uart_nrfx_uarte.c:
```diff
@@ -118,23 +118,23 @@ LOG_MODULE_REGISTER(uart_nrfx_uarte, CONFIG_UART_LOG_LEVEL);
-BUILD_ASSERT(IS_ENABLED(CONFIG_PM_DEVICE_RUNTIME));
-BUILD_ASSERT(IS_ENABLED(CONFIG_CLOCK_CONTROL));
-#define UARTE_ANY_FAST_PD 1
+#if CONFIG_SOC_NRF54H20_CPUAPP || CONFIG_SOC_NRF54H20_CPURAD
```
Could this depend on nrfs services instead?

masz-nordic commented on soc/nordic/common/pinctrl_soc.h:
```diff
@@ -154,6 +152,20 @@ typedef uint32_t pinctrl_soc_pin_t;
+#define NRF_GET_PORT(pincfg) (NRF_GET_PIN(pincfg) >> 5)
+
+/**
+ * @brief Utility macro to obtain port.
```
pin

bjarki-andreasen replied:
changed to `@brief Utility macro to obtain pin relative to port`

masz-nordic reviewed (commented):
Awesome! :)

bjarki-andreasen commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -16,8 +16,10 @@
-#ifdef CONFIG_SOC_NRF54H20_GPD
-#include <nrf/gpd.h>
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_gpio_pad_group)
+#define NRF_GPIO_HAS_PAD_GROUP 1
```
Changed to GPIO_HAS_PAD_GROUP

bjarki-andreasen commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -112,13 +111,105 @@ static const nrf_gpio_pin_drive_t drive_modes[NRF_DRIVE_COUNT] = {
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_pad_group6)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_pad_group7)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_pad_group8)),
+	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(gpio_pad_group9)),
```
Now looks like 
```
#define GPIO_PAD_GROUP_GET_OR_NULL(idx, _) \
	DEVICE_DT_GET_OR_NULL(DT_NODELABEL(_CONCAT(gpio_pad_group, idx)))

static const struct device *const pad_groups[] = {
	LISTIFY(10, GPIO_PAD_GROUP_GET_OR_NULL, (,))
};
```

bjarki-andreasen commented on drivers/pinctrl/pinctrl_nrf.c:
```diff
@@ -472,25 +563,9 @@ int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
-#endif /* CONFIG_SOC_NRF54H20_GPD */
+			/* enable pin */
+			pad_group_request_pin(pin);
+			port_pin_clock_set(pin, NRF_GET_CLOCKPIN_ENABLE(pins[i]));
```
Now clock is enabled by
```
			if (NRF_GET_LP(pins[i]) == NRF_LP_ENABLE) {
				/* disable pin and pin clock */
				pad_group_release_pin(pin);
				port_pin_clock_set(pin, false);
			} else {
				/* configure pin clock */
				port_pin_clock_set(pin, NRF_GET_CLOCKPIN_ENABLE(pins[i]));
			}
```

bjarki-andreasen commented:
Rebased to solve merge conflict

bjarki-andreasen commented:
ping @anangl @nika-nordic

bjarki-andreasen commented on soc/nordic/common/soc_nrf_common.h:
```diff
@@ -268,6 +268,52 @@
+			DT_NODE_EXISTS(DT_PHANDLE(node_id, power_domains)),			\
+			DT_NODE_EXISTS(DT_NODELABEL(gdpwr_fast_active_1))			\
+		),										\
+		(										\
+			IS_EQ(									\
+				DT_DEP_ORD(DT_PHANDLE(node_id, power_domains)),			\
+				DT_DEP_ORD(DT_NODELABEL(gdpwr_fast_active_1))			\
+			)									\
+		),										\
```
This does not work, DT_SAME_NODE uses `==`, it does not evaluate to 0 or 1 :)

anangl replied:
?
https://github.com/zephyrproject-rtos/zephyr/blob/00e217da7d841e1ff163377535f1ac46b701a9cc/include/zephyr/devicetree.h#L692-L693

bjarki-andreasen replied:
oh, this was fixed after I made the commit :D https://github.com/zephyrproject-rtos/zephyr/commit/ba48d83bec18e9f2caf61979aadf9f8537086cee

bjarki-andreasen commented on tests/drivers/spi/spi_controller_peripheral/boards/nrf54h20dk_nrf54h20_cpuapp_fast_spis.overlay:
```diff
@@ -103,6 +103,7 @@
 &spi120 {
 	compatible = "nordic,nrf-spim";
 	status = "okay";
+	def-char = <0x00>;
```
Some driver needs `def-char` to be set, this was an attempt at trying to fix it.

bjarki-andreasen replied:
Issue was with SPIS_NODE macro which checked if an instance was FAST to determine if the nodelabel was spix or spisx, this can't be done since we need to know the nodelabel first, to be able to check if its fast or not, changed to just test if the nodelabel exists:
```
#define SPIS_NODE(idx) \
	COND_CODE_1(DT_NODE_EXISTS(DT_NODELABEL(spis##idx)), (spis##idx), (spi##idx))
```

bjarki-andreasen commented on dts/bindings/gpio/nordic,nrf-gpio-pad-group.yaml:
```diff
@@ -13,9 +13,9 @@ description: |
-  named gpio_pad_group<GPIO number>.
+  controller which manages the pad group, named
+  pad-group. The pad group's nodelabel is named
+  gpio_pad_group<GPIO number>.
```
done

bjarki-andreasen commented on soc/nordic/common/pinctrl_soc.h:
```diff
@@ -154,6 +133,20 @@ typedef uint32_t pinctrl_soc_pin_t;
+ *
+ * @param pincfg Pin configuration bit field.
+ */
+#define NRF_GET_PORT_PIN(pincfg) (NRF_GET_PORT(pincfg) % 32)
```
done

bjarki-andreasen commented on dts/vendor/nordic/nrf54h20.dtsi:
```diff
@@ -177,6 +177,7 @@
 			open-loop-startup-time-us = <200>; /* To be measured */
 			clocks = <&hfxo>, <&lfxo>;
 			clock-names = "hfxo", "lfxo";
+			status = "disabled";
```
done

bjarki-andreasen commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -895,20 +899,13 @@ static int spi_nrfx_deinit(const struct device *dev)
-	CONFIG_SPI_INIT_PRIORITY < CONFIG_CLOCK_CONTROL_NRF_HSFLL_GLOBAL_INIT_PRIORITY
 #define SPIM_INIT_PRIORITY(idx) \
-	COND_CODE_1(SPIM_REQUESTS_CLOCK(SPIM(idx)), \
+	COND_CODE_1(INSTANCE_IS_FAST_PD(_, /*empty*/, idx, _), \
```
done

bjarki-andreasen commented on dts/vendor/nordic/nrf54h20.dtsi:
```diff
@@ -210,9 +213,48 @@
+		current-limit-ua = <10000>;
+		zephyr,pm-device-runtime-auto;
+		#power-domain-cells = <0>;
+		status = "disabled";
```
removed
